#521 - Skill auto reload

augustnmonteiro opened the pull request:
- Checking by modification in skills and reloading

kfezer commented on mycroft/skills/main.py:
```diff
@@ -17,45 +17,99 @@
+            list = filter(lambda x: os.path.isdir(os.path.join(dir, x)),
+                          os.listdir(dir))
+            for skill_folder in list:
+                if not loaded_skills.has_key(skill_folder):
```
Changing this to
```
if not loaded_skills in skill_folder:
```
I think is better. has_key is depricated

kfezer commented on mycroft/skills/main.py:
```diff
@@ -17,45 +17,99 @@
+        del events[event]
+
+
+def load_skills():
```
can we name this something else? We already have load_skills and load_skill in core.py

kfezer commented on mycroft/skills/main.py:
```diff
@@ -17,45 +17,99 @@
+                          os.listdir(dir))
+            for skill_folder in list:
+                if not loaded_skills.has_key(skill_folder):
+                    loaded_skills[skill_folder] = {}
```
I added:
```
logger.debug("Checking for updates to skills...")
```
here to get confirmation it's working. After that line, a confirmation of the new skill will be written by core.py

kfezer commented on mycroft/skills/main.py:
```diff
@@ -17,45 +17,99 @@
+                                                    0) > last_modified_skill:
+                    skill["instance"].shutdown()
+                    clear_skill_events(skill["instance"])
+                    del skill["instance"]
```
after this I added:
```
                    logger.debug("Removed " + skill["instance"])
```
to confirm deletion of skill. Have not tested.

kfezer commented:
Works for me on the RPi

penrods commented:
In watch_skills(), aren't you recursing there at the end?

augustnmonteiro commented:
@penrods yeah, I'm starting a thread and running watch_skill each 2 seconds

penrods commented:
Right, but it is calling itself and never ending.  So every time you call yourself again you are adding a new entry on the call stack.   I.e. you are doing this:

watch_skills()
    watch_skills()
        watch_skills()
            watch_skills()
                ....
Eventually that is going to crash.


Instead, this should be in an infinite loop.  Something like:

    def watch_skills():
        while True:
            # do the check
            sleep(2)

Same behavior, but there is only a single function call sitting on the stack.


Also....add some comments in that code!!!

aatchison commented:
This is very cool, if somewhat verbose at times.
I had a skill that required youtube_dl, so I watched the skills client error for a while. I installed the dependency live and as soon as it could load, it did!

I'm wondering if they will un-load as well?

aatchison commented:
I'm also wondering if we could stop trying eventually if the file system doesn't change and it fails.
`2017-02-23 10:08:45,373 - mycroft.skills.core - INFO - ATTEMPTING TO LOAD SKILL: media
2017-02-23 10:08:45,373 - mycroft.skills.core - WARNING - Module media does not appear to be skill
2017-02-23 10:08:47,383 - mycroft.skills.core - INFO - ATTEMPTING TO LOAD SKILL: media
2017-02-23 10:08:47,383 - mycroft.skills.core - WARNING - Module media does not appear to be skill
2017-02-23 10:08:49,398 - mycroft.skills.core - INFO - ATTEMPTING TO LOAD SKILL: media
2017-02-23 10:08:49,399 - mycroft.skills.core - WARNING - Module media does not appear to be skill
2017-02-23 10:08:51,415 - mycroft.skills.core - INFO - ATTEMPTING TO LOAD SKILL: media
2017-02-23 10:08:51,415 - mycroft.skills.core - WARNING - Module media does not appear to be skill
2017-02-23 10:08:53,426 - mycroft.skills.core - INFO - ATTEMPTING TO LOAD SKILL: media
2017-02-23 10:08:53,427 - mycroft.skills.core - WARNING - Module media does not appear to be skill
2017-02-23 10:08:55,437 - mycroft.skills.core - INFO - ATTEMPTING TO LOAD SKILL: media
2017-02-23 10:08:55,437 - mycroft.skills.core - WARNING - Module media does not appear to be skill
2017-02-23 10:08:57,451 - mycroft.skills.core - INFO - ATTEMPTING TO LOAD SKILL: media
2017-02-23 10:08:57,451 - mycroft.skills.core - WARNING - Module media does not appear to be skill
2017-02-23 10:08:59,466 - mycroft.skills.core - INFO - ATTEMPTING TO LOAD SKILL: media
2017-02-23 10:08:59,466 - mycroft.skills.core - WARNING - Module media does not appear to be skill
2017-02-23 10:09:01,477 - mycroft.skills.core - INFO - ATTEMPTING TO LOAD SKILL: media
2017-02-23 10:09:01,477 - mycroft.skills.core - WARNING - Module media does not appear to be skill
2017-02-23 10:09:03,492 - mycroft.skills.core - INFO - ATTEMPTING TO LOAD SKILL: media
2017-02-23 10:09:03,492 - mycroft.skills.core - WARNING - Module media does not appear to be skill
2017-02-23 10:09:05,506 - mycroft.skills.core - INFO - ATTEMPTING TO LOAD SKILL: media
2017-02-23 10:09:05,507 - mycroft.skills.core - WARNING - Module media does not appear to be skill
2017-02-23 10:09:07,519 - mycroft.skills.core - INFO - ATTEMPTING TO LOAD SKILL: media
2017-02-23 10:09:07,520 - mycroft.skills.core - WARNING - Module media does not appear to be skill
2017-02-23 10:09:09,539 - mycroft.skills.core - INFO - ATTEMPTING TO LOAD SKILL: media
2017-02-23 10:09:09,539 - mycroft.skills.core - WARNING - Module media does not appear to be skill
`

augustnmonteiro commented:
@penrods I didn't remember the call stack, fixed!

@aatchison I changed to load once if filesystem doesn't change it's because before I keep trying load a skill with some error. To unload I'm removing all events attached to the skill and destroy the skill instance

penrods commented:
Still could use some comments, but looks good to me
